lumicks/pylake/tests/test_imaging_confocal/test_kymo_slice_crop.py

```diff
@@ -0,0 +1,273 @@
+    # very last line and less than or equal to the very last timestamp of the infowave created a
+    # dysfunctional Kymo.
+    assert isinstance(kymo[kymo.timestamps[2, -1] :], EmptyKymo)
+    assert isinstance(kymo[f"{(kymo.timestamps[-1, -1] - kymo.start) * 1e-9}s":], EmptyKymo)
+
+
+def test_downsampled_slice(test_kymo):
+    """There was a regression bug that if a Kymo was downsampled and then sliced, it would undo the
+    downsampling. For now, we just flag it as not implemented behaviour."""
+    kymo, _ = test_kymo
+
+    with pytest.raises(NotImplementedError):
+        kymo.downsampled_by(time_factor=2)["1s":"2s"]
+
+
+def test_kymo_crop(cropping_kymo):
+    """Test basic cropping functionality"""
+    kymo, ref = cropping_kymo
+    px_size = ref.metadata.pixelsize_um[0]
+
+    cropped = kymo.crop_by_distance(2 * px_size, 4 * px_size)
+    ref_cropped = ref.image[2:4, :, 0]
+
+    np.testing.assert_allclose(cropped.get_image("red"), ref_cropped)
+    np.testing.assert_allclose(cropped.get_image("rgb")[:, :, 0], ref_cropped)
+    np.testing.assert_allclose(cropped.get_image("rgb")[:, :, 1], np.zeros(ref_cropped.shape))
+    np.testing.assert_allclose(cropped.get_image("green"), np.zeros(ref_cropped.shape))  # missing
+    np.testing.assert_equal(cropped.timestamps, ref.timestamps.data[2:4, :])
+    assert cropped.timestamps.dtype == np.int64
+    np.testing.assert_allclose(cropped.pixelsize_um, kymo.pixelsize_um)
+    np.testing.assert_allclose(cropped.line_time_seconds, kymo.line_time_seconds)
+    np.testing.assert_allclose(cropped.pixels_per_line, 2)
+    np.testing.assert_allclose(cropped._position_offset, 2 * px_size)
+
+    with pytest.raises(ValueError, match="Cropping by negative positions not allowed"):
+        kymo.crop_by_distance(-2 * px_size, px_size)
+
+    with pytest.raises(ValueError, match="Cropping by negative positions not allowed"):
+        kymo.crop_by_distance(px_size, -4 * px_size)
+
+    with pytest.raises(IndexError, match="Cropped image would be empty"):
+        kymo.crop_by_distance(5 * px_size, 2 * px_size)
+
+    with pytest.raises(IndexError, match="Cropped image would be empty"):
+        kymo.crop_by_distance(2 * px_size, 2 * px_size)
+
+    with pytest.raises(IndexError, match="Cropped image would be empty"):
+        kymo.crop_by_distance(20 * px_size, 21 * px_size)
+
+    # Test rounding internally
+    np.testing.assert_allclose(
+        kymo.crop_by_distance(px_size * 1.6, px_size * 1.6).get_image("red"),
+        ref.image[1:2, :, 0],
+    )
+    np.testing.assert_allclose(
+        kymo.crop_by_distance(px_size * 1.6, px_size * 2.1).get_image("red"),
+        ref.image[1:3, :, 0],
+    )
+    np.testing.assert_allclose(
+        kymo.crop_by_distance(px_size * 2.1, px_size * 2.1).get_image("red"),
+        ref.image[2:3, :, 0],
+    )
+
+    # Test cropping in base pairs
+    kymo_bp = kymo.calibrate_to_kbp(1.000)
```

Comment:
I think it might make sense to split the bp ones off into their own test. Also because you essentially redefine `px_size` here.